File: supabase/tests/concurrency/tracking.py
#!/usr/bin/env python3
"""Exercise Cycle 6 lock ordering with two real PostgreSQL sessions.

The harness creates the existing Cycle 4 fixture inside the isolated
``vango_cycle_6`` database, commits only that disposable fixture, and removes
it in a finally block.  It never accepts a remote host or the shared
``postgres`` database.
"""

from __future__ import annotations

import os
import re
import select
import shutil
import subprocess
import tempfile
import time
from pathlib import Path
from urllib.parse import parse_qs, urlparse


ROOT = Path(__file__).resolve().parents[3]
PSQL_DEFAULT = "/opt/homebrew/opt/libpq/bin/psql"
LOCAL_HOSTS = {"127.0.0.1", "localhost", "::1"}
DATABASE = "vango_cycle_6"
MARKER_RE = re.compile(r"^VANGO_TRACKING_([A-Z_]+)=([0-9a-fA-F-]+)$")
UUID_RE = re.compile(r"^[0-9a-fA-F]{8}(?:-[0-9a-fA-F]{4}){3}-[0-9a-fA-F]{12}$")
FIXTURE_USER_IDS = (
    "40000000-0000-0000-0000-000000000001",
    "40000000-0000-0000-0000-000000000002",
    "40000000-0000-0000-0000-000000000003",
    "40000000-0000-0000-0000-000000000004",
    "40000000-0000-0000-0000-000000000005",
    "60000000-0000-0000-0000-000000000001",
    "60000000-0000-0000-0000-000000000002",
    "60000000-0000-0000-0000-000000000003",
    "60000000-0000-0000-0000-000000000004",
    "60000000-0000-0000-0000-000000000005",
)
FIXTURE_FLEET_IDS = (
    "41000000-0000-0000-0000-000000000001",
    "41000000-0000-0000-0000-000000000002",
)
FIXTURE_SCHOOL_ID = "65000000-0000-0000-0000-000000000001"


class TrackingRaceError(AssertionError):
    """A concurrency invariant was not demonstrated."""


def database_url() -> str:
    value = os.environ.get(
        "VANGO_TEST_DATABASE_URL",
        "postgresql://supabase_admin:postgres@127.0.0.1:54322/vango_cycle_6",
    )
    parsed = urlparse(value)
    if parsed.scheme not in {"postgres", "postgresql"}:
        raise TrackingRaceError("database URL must use postgres or postgresql")
    if parsed.hostname not in LOCAL_HOSTS:
        raise TrackingRaceError("refusing a non-loopback database host")
    try:
        parsed_port = parsed.port
    except ValueError as error:
        raise TrackingRaceError("database URL has an invalid port") from error
    if parsed_port != 54322:
        raise TrackingRaceError("database URL must use the local Supabase port 54322")
    query_keys = {key.lower() for key in parse_qs(parsed.query, keep_blank_values=True)}
    if query_keys.intersection({"host", "hostaddr", "service", "servicefile", "port", "dbname"}):
        raise TrackingRaceError("database URL cannot override its local connection target")
    if any(
        os.environ.get(key)
        for key in (
            "PGHOST",
            "PGHOSTADDR",
            "PGPORT",
            "PGDATABASE",
            "PGSERVICE",
            "PGSERVICEFILE",
        )
    ):
        raise TrackingRaceError("libpq environment cannot override the local connection target")
    database = parsed.path.removeprefix("/")
    if database != DATABASE:
        raise TrackingRaceError(
            f"refusing a database other than {DATABASE} (got {database!r})"
        )
    return value


def psql_path() -> str:
    requested = os.environ.get("VANGO_PSQL", PSQL_DEFAULT)
    if os.path.isabs(requested):
        if not os.access(requested, os.X_OK):
            raise TrackingRaceError(f"psql is not executable: {requested}")
        return requested
    resolved = shutil.which(requested)
    if not resolved:
        raise TrackingRaceError("psql was not found; set VANGO_PSQL")
    return resolved


def command_args(psql: str, dsn: str, *extra: str) -> list[str]:
    return [psql, dsn, "-X", "-v", "ON_ERROR_STOP=1", *extra]


def run_sql(psql: str, dsn: str, sql: str) -> str:
    completed = subprocess.run(
        command_args(psql, dsn, "-A", "-t", "-P", "pager=off", "-c", sql),
        cwd=ROOT,
        env=os.environ.copy(),
        text=True,
        capture_output=True,
        check=False,
        timeout=20,
    )
    if completed.returncode != 0:
        raise TrackingRaceError(
            f"observation query failed with exit {completed.returncode}: "
            f"{completed.stderr.strip()}"
        )
    return completed.stdout.strip()


def assert_fixture_ids_free(psql: str, dsn: str) -> None:
    user_ids = ", ".join(f"'{value}'::uuid" for value in FIXTURE_USER_IDS)
    fleet_ids = ", ".join(f"'{value}'::uuid" for value in FIXTURE_FLEET_IDS)
    query = f"""
select count(*)
from (
  select id from auth.users where id in ({user_ids})
  union all
  select id from public.fleets where id in ({fleet_ids})
  union all
  select id from public.schools where id = '{FIXTURE_SCHOOL_ID}'::uuid
) occupied;
"""
    occupied = run_sql(psql, dsn, query)
    if occupied != "0":
        raise TrackingRaceError(
            "tracking fixture IDs are already occupied; refusing to overwrite them"
        )


def setup_fixture(psql: str, dsn: str) -> tuple[dict[str, str], Path]:
    setup_sql = r"""
begin;
\ir __HELPERS__
\ir __PLANNING__
\ir __OPERATIONS__
\ir __TRACKING__
select pg_temp.seed_tracking();
select set_config(
  'request.jwt.claims',
  '{"sub":"40000000-0000-0000-0000-000000000001","role":"authenticated"}',
  true
);
select public.request_trip_route_calculation(
  (select id from operation_ids where kind = 'return')
) as route_revision \gset
select id as active_trip
from tracking_ids where kind = 'trip' \gset
select id as active_assignment
from tracking_ids where kind = 'assignment' \gset
select id as active_student
from tracking_ids where kind = 'student' \gset
select id as active_fleet
from tracking_ids where kind = 'fleet' \gset
select id as return_trip
from operation_ids where kind = 'return' \gset
select :'route_revision' as return_route_revision \gset
select 'VANGO_TRACKING_TRIP=' || :'active_trip';
select 'VANGO_TRACKING_ASSIGNMENT=' || :'active_assignment';
select 'VANGO_TRACKING_STUDENT=' || :'active_student';
select 'VANGO_TRACKING_FLEET=' || :'active_fleet';
select 'VANGO_TRACKING_RETURN_TRIP=' || :'return_trip';
select 'VANGO_TRACKING_ROUTE_REVISION=' || :'return_route_revision';
commit;
"""
    for placeholder, path in (
        ("__HELPERS__", ROOT / "supabase/tests/_helpers.psql"),
        ("__PLANNING__", ROOT / "supabase/tests/_planning.psql"),
        ("__OPERATIONS__", ROOT / "supabase/tests/_operations.psql"),
        ("__TRACKING__", ROOT / "supabase/tests/_tracking.psql"),
    ):
        setup_sql = setup_sql.replace(placeholder, str(path))
    with tempfile.NamedTemporaryFile(
        mode="w", suffix=".psql", prefix="vango-cycle6-setup-", delete=False
    ) as handle:
        handle.write(setup_sql)
        setup_path = Path(handle.name)
    committed = False
    try:
        completed = subprocess.run(
            command_args(psql, dsn, "-A", "-t", "-P", "pager=off", "-f", str(setup_path)),
            cwd=ROOT,
            env=os.environ.copy(),
            text=True,
            capture_output=True,
            check=False,
            timeout=30,
        )
        if completed.returncode != 0:
            raise TrackingRaceError(
                f"tracking fixture setup failed: {completed.stderr.strip()}"
            )
        committed = True
        values: dict[str, str] = {}
        for line in completed.stdout.splitlines():
            match = MARKER_RE.fullmatch(line.strip())
            if match:
                key, value = match.groups()
                if key == "ROUTE_REVISION":
                    if not value.isdigit() or int(value) < 1:
                        raise TrackingRaceError("setup returned an invalid route revision")
                elif not UUID_RE.fullmatch(value):
                    raise TrackingRaceError(f"setup returned an invalid UUID for {key}")
                values[key] = value.lower()
        required = {
            "TRIP", "ASSIGNMENT", "STUDENT", "FLEET", "RETURN_TRIP", "ROUTE_REVISION"
        }
        missing = required - values.keys()
        if missing:
            raise TrackingRaceError(f"setup did not return markers: {sorted(missing)}")
        return values, setup_path
    except Exception:
        if committed:
            cleanup_fixture(psql, dsn, setup_path)
        else:
            setup_path.unlink(missing_ok=True)
        raise


def cleanup_fixture(psql: str, dsn: str, setup_path: Path) -> None:
    setup_path.unlink(missing_ok=True)
    cleanup_sql = """
begin;
delete from private.notification_manual_commands
where fleet_id in ('41000000-0000-0000-0000-000000000001', '41000000-0000-0000-0000-000000000002');
delete from private.trip_location_summaries
where fleet_id in ('41000000-0000-0000-0000-000000000001', '41000000-0000-0000-0000-000000000002');
delete from public.trip_location_points
where fleet_id in ('41000000-0000-0000-0000-000000000001', '41000000-0000-0000-0000-000000000002');
delete from public.trip_location_receipts
where fleet_id in ('41000000-0000-0000-0000-000000000001', '41000000-0000-0000-0000-000000000002');
delete from public.trip_current_locations
where fleet_id in ('41000000-0000-0000-0000-000000000001', '41000000-0000-0000-0000-000000000002');
delete from public.trip_route_calculations
where fleet_id in ('41000000-0000-0000-0000-000000000001', '41000000-0000-0000-0000-000000000002');
delete from public.notifications
where fleet_id in ('41000000-0000-0000-0000-000000000001', '41000000-0000-0000-0000-000000000002');
commit;
"""
    # The C6 rows are deleted first because the existing C4 fixture cleanup
    # removes assignments and trips.  Reuse that narrow, fixed-fleet cleanup
    # instead of duplicating its full FK order here.
    cleanup_sql += "\n" + (ROOT / "supabase/tests/concurrency/operations_cleanup.psql").read_text()
    completed = subprocess.run(
        command_args(psql, dsn, "-A", "-t", "-P", "pager=off", "-c", cleanup_sql),
        cwd=ROOT,
        env=os.environ.copy(),
        text=True,
        capture_output=True,
        check=False,
        timeout=30,
    )
    if completed.returncode != 0:
        raise TrackingRaceError(f"tracking C6 cleanup failed: {completed.stderr.strip()}")


def session(psql: str, dsn: str, application_name: str) -> subprocess.Popen[str]:
    environment = os.environ.copy()
    environment["PGAPPNAME"] = application_name
    return subprocess.Popen(
        command_args(psql, dsn, "-A", "-t", "-P", "pager=off"),
        cwd=ROOT,
        env=environment,
        stdin=subprocess.PIPE,
        stdout=subprocess.PIPE,
        stderr=subprocess.PIPE,
        text=True,
        bufsize=1,
    )


def send(process: subprocess.Popen[str], sql: str) -> None:
    if process.stdin is None:
        raise TrackingRaceError("psql stdin is unavailable")
    process.stdin.write(sql)
    process.stdin.flush()


def wait_for_marker(process: subprocess.Popen[str], marker: str) -> None:
    if process.stdout is None:
        raise TrackingRaceError("psql stdout is unavailable")
    deadline = time.monotonic() + 8
    output = b""
    while time.monotonic() < deadline:
        remaining = max(0.0, deadline - time.monotonic())
        ready, _, _ = select.select([process.stdout], [], [], remaining)
        if not ready:
            break
        chunk = os.read(process.stdout.fileno(), 4096)
        if not chunk:
            raise TrackingRaceError(f"session ended before {marker}")
        output += chunk
        if marker.encode() in output.splitlines():
            return
    raise TrackingRaceError(f"session did not publish {marker}")


def wait_for_advisory_wait(psql: str, dsn: str, application_name: str) -> None:
    deadline = time.monotonic() + 8
    query = (
        "select count(*) from pg_stat_activity "
        f"where application_name = '{application_name}' "
        "and wait_event = 'advisory'"
    )
    while time.monotonic() < deadline:
        if run_sql(psql, dsn, query) == "1":
            return
        time.sleep(0.05)
    raise TrackingRaceError(f"{application_name} did not wait on planning lock")


def finish_process(process: subprocess.Popen[str], timeout: float = 12) -> tuple[int, str, str]:
    if process.stdin is not None and not process.stdin.closed:
        process.stdin.close()
    try:
        process.wait(timeout=timeout)
    except subprocess.TimeoutExpired as error:
        process.kill()
        process.wait(timeout=3)
        raise TrackingRaceError("concurrency session timed out") from error
    stdout = process.stdout.read() if process.stdout is not None else ""
    stderr = process.stderr.read() if process.stderr is not None else ""
    return process.returncode, stdout, stderr


def assert_revocation_race(psql: str, dsn: str, ids: dict[str, str]) -> None:
    trip = ids["TRIP"]
    assignment = ids["ASSIGNMENT"]
    driver = "40000000-0000-0000-0000-000000000003"
    first = session(psql, dsn, "vango-cycle6-revocation-first")
    second = session(psql, dsn, "vango-cycle6-ingest-second")
    try:
        send(
            first,
            "begin;\n"
            "select private.lock_planning();\n"
            f"update public.trip_assignments set valid_until = clock_timestamp() "
            f"where id = '{assignment}' and valid_until is null;\n"
            "select 'VANGO_REVOKED';\n",
        )
        wait_for_marker(first, "VANGO_REVOKED")
        send(
            second,
            "begin;\n"
            f"select set_config('request.jwt.claims', '{{\"sub\":\"{driver}\","
            "\"role\":\"authenticated\"}', false);\n"
            f"select public.ingest_trip_locations('{trip}'::uuid, '{assignment}'::uuid, "
            "jsonb_build_array(jsonb_build_object('sequence', 990001, "
            "'captured_at', clock_timestamp(), 'latitude', -23.5, 'longitude', -46.6, "
            "'accuracy', 5)), true);\n"
            "commit;\n",
        )
        wait_for_advisory_wait(psql, dsn, "vango-cycle6-ingest-second")
        send(first, "commit;\n")
        first_result = finish_process(first)
        second_result = finish_process(second)
        if first_result[0] != 0:
            raise TrackingRaceError(f"revocation session failed: {first_result[2]}")
        if second_result[0] == 0 or '"code": "invalid_transition"' not in second_result[2] or "Live GPS requires an active trip assignment" not in second_result[2]:
            raise TrackingRaceError(
                "ingest after committed assignment revocation did not fail closed: "
                + second_result[2]
            )
    finally:
        for process in (first, second):
            if process.poll() is None:
                process.kill()
                process.wait(timeout=3)


def assert_route_revision_race(psql: str, dsn: str, ids: dict[str, str]) -> None:
    trip = ids["RETURN_TRIP"]
    revision = ids["ROUTE_REVISION"]
    first = session(psql, dsn, "vango-cycle6-route-edit-first")
    second = session(psql, dsn, "vango-cycle6-route-apply-second")
    try:
        send(
            first,
            "begin;\n"
            "select private.lock_planning();\n"
            f"update public.trip_stops set address_snapshot = '{{\"label\":\"race\"}}'::jsonb "
            f"where trip_id = '{trip}'::uuid and reached_at is null "
            "and id = (select id from public.trip_stops where trip_id = '"
            f"{trip}'::uuid and reached_at is null order by position limit 1);\n"
            "select 'VANGO_ROUTE_EDITED';\n",
        )
        wait_for_marker(first, "VANGO_ROUTE_EDITED")
        send(
            second,
            f"select public.apply_trip_route_result('{trip}'::uuid, {revision}::bigint, "
            "'{}'::jsonb);\n",
        )
        wait_for_advisory_wait(psql, dsn, "vango-cycle6-route-apply-second")
        send(first, "commit;\n")
        first_result = finish_process(first)
        second_result = finish_process(second)
        if first_result[0] != 0:
            raise TrackingRaceError(f"route edit session failed: {first_result[2]}")
        if second_result[0] != 0 or "superseded" not in second_result[1]:
            raise TrackingRaceError(
                "stale route result was not superseded after concurrent edit: "
                + second_result[1] + second_result[2]
            )
        state = run_sql(
            psql,
            dsn,
            f"select status from public.trip_route_calculations "
            f"where trip_id = '{trip}'::uuid and revision = {revision};",
        )
        if state != "superseded":
            raise TrackingRaceError(f"stale calculation state was {state!r}")
    finally:
        for process in (first, second):
            if process.poll() is None:
                process.kill()
                process.wait(timeout=3)


def main() -> int:
    dsn = database_url()
    psql = psql_path()
    current_database = run_sql(psql, dsn, "select current_database();")
    if current_database != DATABASE:
        raise TrackingRaceError(f"refusing to run outside {DATABASE} (got {current_database!r})")
    assert_fixture_ids_free(psql, dsn)
    ids: dict[str, str] | None = None
    setup_path: Path | None = None
    try:
        ids, setup_path = setup_fixture(psql, dsn)
        assert_revocation_race(psql, dsn, ids)
        assert_route_revision_race(psql, dsn, ids)
        print("tracking concurrency: revocation×ingest and route revision×apply PASS")
        return 0
    finally:
        if setup_path is not None:
            cleanup_fixture(psql, dsn, setup_path)


if __name__ == "__main__":
    try:
        raise SystemExit(main())
    except TrackingRaceError as error:
        raise SystemExit(str(error)) from error

File: supabase/tests/concurrency/tracking.ts
// Actual local Realtime sockets, including channels opened before revocation.
// Run: deno run --allow-run --allow-read --allow-env --allow-net tracking.ts
const root = new URL("../../../", import.meta.url);
const psql = Deno.env.get("VANGO_PSQL") ?? "/opt/homebrew/opt/libpq/bin/psql";
const sqlEnv = {
  PGHOST: "127.0.0.1",
  PGPORT: "54322",
  PGUSER: "postgres",
  PGDATABASE: "postgres",
  PGPASSWORD: "postgres",
  PGOPTIONS: "-c statement_timeout=20000 -c lock_timeout=15000",
};
for (const key of ["PGHOSTADDR", "PGSERVICE", "PGSERVICEFILE"]) {
  if (Deno.env.get(key)) {
    throw new Error(`${key} must be unset for local-only tracking tests`);
  }
}
function assert(condition: unknown, message: string): asserts condition {
  if (!condition) throw new Error(message);
}
async function command(
  program: string,
  args: string[],
  input?: string,
): Promise<string> {
  const child = new Deno.Command(program, {
    args,
    cwd: root,
    env: sqlEnv,
    stdin: input === undefined ? "null" : "piped",
    stdout: "piped",
    stderr: "piped",
  }).spawn();
  if (input !== undefined) {
    const writer = child.stdin.getWriter();
    await writer.write(new TextEncoder().encode(input));
    await writer.close();
  }
  const result = await child.output();
  assert(
    result.success,
    program === psql
      ? `Local fixture SQL failed: ${new TextDecoder().decode(result.stderr)}`
      : `${program} failed; credentials and private output suppressed`,
  );
  return new TextDecoder().decode(result.stdout);
}
function sql(text: string): Promise<string> {
  return command(psql, ["-X", "-qAt", "-v", "ON_ERROR_STOP=1"], text);
}
const status = JSON.parse(await command("supabase", ["status", "-o", "json"]));
const api = new URL(status.API_URL);
assert(
  ["127.0.0.1", "localhost", "[::1]"].includes(api.hostname) &&
    api.port === "54321",
  "Refusing non-local API",
);
const fixtureFleet = "41000000-0000-0000-0000-000000000001";
assert(
  (await sql(
    `select count(*) from auth.users where id::text like '40000000-%' or id::text like '60000000-%';`,
  )).trim() === "0",
  "Dedicated fixtures already exist; refusing to overwrite",
);
const encoder = new TextEncoder();
const base64 = (bytes: Uint8Array) =>
  btoa(String.fromCharCode(...bytes)).replaceAll("+", "-").replaceAll("/", "_")
    .replaceAll("=", "");
async function token(user: string): Promise<string> {
  const now = Math.floor(Date.now() / 1000);
  const data = `${
    base64(encoder.encode(JSON.stringify({ alg: "HS256", typ: "JWT" })))
  }.${
    base64(
      encoder.encode(
        JSON.stringify({
          sub: user,
          role: "authenticated",
          aud: "authenticated",
          iat: now,
          exp: now + 3600,
        }),
      ),
    )
  }`;
  const key = await crypto.subtle.importKey(
    "raw",
    encoder.encode(status.JWT_SECRET),
    { name: "HMAC", hash: "SHA-256" },
    false,
    ["sign"],
  );
  return `${data}.${
    base64(
      new Uint8Array(
        await crypto.subtle.sign("HMAC", key, encoder.encode(data)),
      ),
    )
  }`;
}
const sockets: WebSocket[] = [];
type Channel = { socket: WebSocket; messages: Record<string, unknown>[] };
async function join(
  topic: string,
  jwt: string,
  allowed = true,
  privateChannel = true,
): Promise<Channel> {
  const url = new URL("/realtime/v1/websocket", api);
  url.protocol = "ws:";
  url.searchParams.set("apikey", status.ANON_KEY);
  url.searchParams.set("vsn", "1.0.0");
  const socket = new WebSocket(url);
  sockets.push(socket);
  const messages: Record<string, unknown>[] = [];
  await new Promise<void>((resolve, reject) => {
    const timer = setTimeout(
      () => reject(new Error("Realtime join timed out")),
      8000,
    );
    socket.onopen = () =>
      socket.send(JSON.stringify({
        topic: `realtime:${topic}`,
        event: "phx_join",
        payload: {
          config: {
            broadcast: { ack: true, self: false },
            presence: { enabled: false },
            postgres_changes: [],
            private: privateChannel,
          },
          access_token: jwt,
        },
        ref: "1",
        join_ref: "1",
      }));
    socket.onerror = () => {
      clearTimeout(timer);
      reject(new Error("Realtime socket failed"));
    };
    socket.onmessage = (event) => {
      const message = JSON.parse(event.data);
      if (
        message.event === "broadcast" && message.payload?.event === "location"
      ) messages.push(message.payload.payload);
      if (message.event === "phx_reply" && message.ref === "1") {
        clearTimeout(timer);
        if ((message.payload.status === "ok") !== allowed) {
          reject(
            new Error(
              allowed
                ? `Authorized private join denied: ${
                  JSON.stringify(message.payload.response)
                }`
                : "Revoked private join accepted",
            ),
          );
        } else resolve();
      }
    };
  });
  return { socket, messages };
}
const delay = (ms: number) => new Promise((resolve) => setTimeout(resolve, ms));
let ownsFixture = false;
try {
  const setup = await command(psql, [
    "-X",
    "-qAt",
    "-v",
    "ON_ERROR_STOP=1",
    "-f",
    new URL("supabase/tests/concurrency/tracking_setup.psql", root).pathname,
  ]);
  ownsFixture = true;
  const line = setup.split("\n").find((line) =>
    line.startsWith("VANGO_TRACKING=")
  );
  assert(line, "Fixture IDs missing");
  const ids: Record<string, string> = JSON.parse(
    line.slice("VANGO_TRACKING=".length),
  );
  assert(
    Object.values(ids).every((id) => /^[0-9a-f-]{36}$/.test(id)),
    "Invalid fixture identifier",
  );
  const owner1 = await token("40000000-0000-0000-0000-000000000001");
  const owner2 = await token("40000000-0000-0000-0000-000000000002");
  const guardian = await token("60000000-0000-0000-0000-000000000001");
  const driver = await token(ids.driver);
  const topic = () =>
    sql(
      `select 'trip:'||id||':v'||broadcast_epoch from public.trips where id='${ids.trip}';`,
    ).then((s) => s.trim());
  const original = await topic();
  const oldOwner = await join(original, owner2);
  const oldGuardian = await join(original, guardian);
  const oldDriver = await join(original, driver);
  const publicChannel = await join(original, status.ANON_KEY, true, false);
  oldOwner.socket.send(
    JSON.stringify({
      topic: `realtime:${original}`,
      event: "broadcast",
      payload: {
        type: "broadcast",
        event: "location",
        payload: { sequence: 777 },
      },
      ref: "fake",
      join_ref: "1",
    }),
  );
  await join(
    original,
    await token("40000000-0000-0000-0000-000000000005"),
    false,
  );
  let sequence = 0;
  const publish = async (
    jwt: string,
    assignment: string,
    positive: Channel,
    negatives: Channel[] = [],
  ): Promise<void> => {
    await delay(1100);
    sequence++;
    const response = await fetch(
      new URL("/rest/v1/rpc/ingest_trip_locations", api),
      {
        method: "POST",
        signal: AbortSignal.timeout(8000),
        headers: {
          apikey: status.ANON_KEY,
          Authorization: `Bearer ${jwt}`,
          "Content-Type": "application/json",
        },
        body: JSON.stringify({
          p_trip_id: ids.trip,
          p_assignment_id: assignment,
          p_points: [{
            sequence,
            captured_at: new Date().toISOString(),
            latitude: -23.55,
            longitude: -46.63,
            accuracy: 5,
          }],
          p_live: true,
        }),
      },
    );
    assert(response.ok, `GPS RPC failed (${response.status})`);
    const deadline = Date.now() + 5000;
    while (
      !positive.messages.some((message) => message.sequence === sequence) &&
      Date.now() < deadline
    ) await delay(25);
    assert(
      positive.messages.some((message) => message.sequence === sequence),
      "Authorized current topic did not receive GPS",
    );
    await delay(300);
    assert(
      negatives.every((channel) =>
        !channel.messages.some((message) => message.sequence === sequence)
      ),
      "An old opened topic received post-revocation GPS",
    );
  };
  await publish(driver, ids.assignment, oldOwner, [publicChannel]);
  assert(
    !oldGuardian.messages.some((message) => message.sequence === 777),
    "Client published an unauthorized location",
  );
  assert(
    oldGuardian.messages.length > 0 && oldDriver.messages.length > 0,
    "Initial authorized participants did not receive GPS",
  );
  await sql(
    `begin; select set_config('request.jwt.claims','{"sub":"40000000-0000-0000-0000-000000000001","role":"authenticated"}',true); select public.set_fleet_membership_status('42000000-0000-0000-0000-000000000002','suspended'); commit;`,
  );
  const afterOwner = await topic();
  assert(afterOwner !== original, "Owner membership loss did not rotate epoch");
  const currentOwner = await join(afterOwner, owner1);
  await join(afterOwner, owner2, false);
  await publish(driver, ids.assignment, currentOwner, [
    oldOwner,
    oldGuardian,
    oldDriver,
  ]);
  await sql(
    `update public.trip_passengers set operation_status='absent' where trip_id='${ids.trip}' and student_id='${ids.student}';`,
  );
  const afterAbsent = await topic();
  const remainingDependent = await join(afterAbsent, guardian);
  await publish(driver, ids.assignment, remainingDependent, [
    oldGuardian,
    currentOwner,
  ]);
  await sql(
    `update public.student_guardians set status='removed',removed_at=clock_timestamp() where student_id='${
      ids["second-dependent"]
    }' and guardian_user_id='60000000-0000-0000-0000-000000000001';`,
  );
  const afterUnlink = await topic();
  await join(afterUnlink, guardian, false);
  const afterPassenger = await join(afterUnlink, owner1);
  await publish(driver, ids.assignment, afterPassenger, [
    remainingDependent,
    oldGuardian,
  ]);
  // Substitute the current assignment; the old driver remains a fleet driver.
  await sql(
    `begin; insert into public.fleet_membership_roles(membership_id,role) select id,'driver' from public.fleet_memberships where fleet_id='${fixtureFleet}' and user_id='40000000-0000-0000-0000-000000000001' on conflict do nothing; update public.trip_assignments set valid_until=clock_timestamp() where id='${ids.assignment}'; update public.trips set driver_user_id='40000000-0000-0000-0000-000000000001' where id='${ids.trip}'; insert into public.trip_assignments(fleet_id,trip_id,van_id,driver_user_id) select fleet_id,id,van_id,driver_user_id from public.trips where id='${ids.trip}'; commit;`,
  );
  const newAssignment = (await sql(
    `select id from public.trip_assignments where trip_id='${ids.trip}' and valid_until is null;`,
  )).trim();
  const afterSwap = await topic();
  await join(afterSwap, driver, false);
  const newDriver = await join(afterSwap, owner1);
  await publish(owner1, newAssignment, newDriver, [oldDriver, afterPassenger]);
  const adultToken = await token("60000000-0000-0000-0000-000000000003");
  const adultChannel = await join(await topic(), adultToken);
  await sql(
    `delete from public.fleet_membership_roles where role='student' and membership_id in (select id from public.fleet_memberships where fleet_id='${fixtureFleet}' and user_id='60000000-0000-0000-0000-000000000003');`,
  );
  const afterAdult = await topic();
  await join(afterAdult, adultToken, false);
  const ownerAfterAdult = await join(afterAdult, owner1);
  await publish(owner1, newAssignment, ownerAfterAdult, [
    adultChannel,
    newDriver,
  ]);
  await sql(
    `update public.fleet_memberships set status='active',suspended_at=null where fleet_id='${fixtureFleet}' and user_id='40000000-0000-0000-0000-000000000002';`,
  );
  const ownerBeforeRoleLoss = await join(await topic(), owner2);
  await sql(
    `delete from public.fleet_membership_roles where role='owner' and membership_id='42000000-0000-0000-0000-000000000002';`,
  );
  await join(await topic(), owner2, false);
  const finalOwner = await join(await topic(), owner1);
  await publish(owner1, newAssignment, finalOwner, [
    ownerBeforeRoleLoss,
    ownerAfterAdult,
  ]);
  console.log(
    "Realtime WebSocket PASS: private delivery, foreign/public/client-write denial, owner suspension/role removal, remaining dependent, guardian unlink, adult role removal and driver substitution; old opened topics stayed silent.",
  );
} catch (error) {
  console.error(
    error instanceof Error ? error.message : "Tracking test failed",
  );
  throw error;
} finally {
  for (const socket of sockets) socket.close();
  if (ownsFixture) {
    const cleanup = await Deno.readTextFile(
      new URL("supabase/tests/concurrency/operations_cleanup.psql", root),
    );
    await sql(cleanup);
    assert(
      (await sql(
        `select count(*) from public.fleets where id='${fixtureFleet}';`,
      )).trim() === "0",
      "Tracking fixture cleanup failed",
    );
    console.log("Realtime fixture cleanup PASS");
  }
}
